File: T5/AvdeevaAN/perf1.py
#!/usr/bin/python

import random
import time

def radixSort(aList):
    Radix = 10
    maxLength = False
    tmp, placement = -1, 1
    
    while not maxLength:
        maxLength = True
        buckets = [list() for _ in range(Radix)]
        for i in aList:
            tmp = i // placement
            buckets[tmp % Radix].append(i)
            if maxLength and tmp > 0:
                maxLength = False
        a = 0
        for b in range(Radix):
            buck = buckets[b]
            for i in buck:
                aList[a] = i
                a += 1
        placement *= Radix
    return aList

def test(size):
    arr = []
    for i in range(size):
        arr.append(random.randint(1,size))

    start_time = time.time()
    print "Sorted array sizeof", size
    radixSort(arr);
    print("--- %s seconds ---" % (time.time() - start_time))

test(10000000)

File: T5/AvdeevaAN/perf1.cpp
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include<vector>

using namespace std;

void radixSort(vector<int>& a)
{
    bool maxLength = false;
    int placement = 1;
    while(!maxLength)
    {
        maxLength = true;
        vector<vector<int>> buckets;
        buckets.resize(10);

        for(size_t i = 0; i < a.size(); i++)
        {
            int tmp = a[i] / placement;
            buckets[tmp % 10].push_back(a[i]);
            if(maxLength && tmp > 0) maxLength = false;
        }

        size_t idx = 0;
        for(size_t b = 0; b < 10; b++)
        {
            for(auto& it : buckets[b])
            {
                a[idx++] = it;
            }
        }
        placement *= 10;
    }
}

void test(size_t size)
{
    vector<int> a;
    a.resize(size);
    for(size_t i = 0; i < size; i++)
    {
        a[i] = rand() % size;
    }

    time_t t1 = time(0);
    radixSort(a);
    time_t t2 = time(0);
    printf("Array size %d sorted in %d sec\n", size, t2-t1);
}

int main()
{
    test(10000000); //0 sec
}